spark/v3.2/spark-extensions/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -46,6 +48,15 @@ case class MergeRowsExec(
     output: Seq[Attribute],
     child: SparkPlan) extends UnaryExecNode {
 
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    if (performCardinalityCheck) {
+      // request a local sort by the row ID attrs to co-locate matches for the same target row
+      Seq(rowIdAttrs.map(attr => SortOrder(attr, Ascending)))
```

Comment:
Instead of requesting a sort, we could keep track of all seen row IDs in a task (we keep only the last one right now). That will potentially require more memory.

Reply:
Feels like this is all leaning heavily on implementation details to actually do our check. While I think this fix is fine for now, maybe we should consider actually doing a group by the cardinality check or a custom aggregation which just throws an error on duplicate records?

Reply:
We used to rely on a separate query and an accumulator, which was a source of issues too. I think we will have more options when we move this to Spark (e.g. we will be able to insert a rule at the query prep stage when we already know the physical plan).

Reply:
Yeah our old approach was also not great, I guess if we have things built into spark it will be a lot safer